storage/s3_test.go

```diff
@@ -398,3 +406,178 @@ func TestS3Retry(t *testing.T) {
 		})
 	}
 }
+
+// Credit to aws-sdk-go
+func val(i interface{}, s string) interface{} {
+	v, err := awsutil.ValuesAtPath(i, s)
+	if err != nil || len(v) == 0 {
+		return nil
+	}
+	if _, ok := v[0].(io.Reader); ok {
+		return v[0]
+	}
+
+	if rv := reflect.ValueOf(v[0]); rv.Kind() == reflect.Ptr {
+		return rv.Elem().Interface()
+	}
+
+	return v[0]
+}
+
+func TestS3CopyEncryptionRequest(t *testing.T) {
+	testcases := []struct {
+		name     string
+		sse      string
+		sseKeyId string
+
+		expectedSSE      string
+		expectedSSEKeyId string
+	}{
+		{
+			name: "no encryption, by default",
+		},
+		{
+			name: "aws:kms encryption with server side generated keys",
+			sse:  "aws:kms",
+
+			expectedSSE: "aws:kms",
+		},
+		{
+			name:     "aws:kms encryption with user provided key",
+			sse:      "aws:kms",
+			sseKeyId: "sdkjn12SDdci#@#EFRFERTqW/ke",
+
+			expectedSSE:      "aws:kms",
+			expectedSSEKeyId: "sdkjn12SDdci#@#EFRFERTqW/ke",
+		},
+		{
+			name:     "provide key without encryption flag, shall be ignored",
+			sseKeyId: "1234567890",
+		},
+	}
+	u, err := url.New("s3://bucket/key")
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+
+			mockApi := s3.New(unit.Session)
+
+			mockApi.Handlers.Unmarshal.Clear()
+			mockApi.Handlers.UnmarshalMeta.Clear()
+			mockApi.Handlers.UnmarshalError.Clear()
+			mockApi.Handlers.Send.Clear()
+
+			mockApi.Handlers.Send.PushBack(func(r *request.Request) {
+
+				r.HTTPResponse = &http.Response{
+					StatusCode: http.StatusOK,
+					Body:       ioutil.NopCloser(strings.NewReader("")),
+				}
+
+				params := r.Params
+				sse := val(params, "ServerSideEncryption")
+				key := val(params, "SSEKMSKeyId")
+
+				if !(sse == nil && tc.expectedSSE == "") {
+					assert.Assert(t, cmp.Equal(sse, tc.expectedSSE))
```

Comment:
```suggestion
					assert.Equal(t, sse, tc.expectedSSE)
```
This produces more informative error messages when tests fail.

Same for other `assert.Assert(t, cmp.Equal(.....))` usages for strings.